moon/controller.py

```diff
@@ -479,6 +483,48 @@ def on_scan(self, data):
         before_robot.sort();
         self.scan_distance_to_obstacle = median(before_robot[:self.scan_min_window])
 
+        # unless we are already inside unsafe area, augment nearest obstacles with virtual fence values
+        if self.unsafe_location is not None and not self.unsafe_location.contains(Point(self.xyz[0], self.xyz[1])):
+            # calculate position in front of and to the left and right of the robot in the direction of its heading
+            v = np.asmatrix(np.asarray([self.xyz[0], self.xyz[1], 1]))
+            c, s = np.cos(self.steering_angle + self.yaw), np.sin(self.steering_angle + self.yaw)
+            R = np.asmatrix(np.array(((c, -s, 0), (s, c, 0), (0, 0, 1))))
+
+            T = np.asmatrix(np.array(((1, 0, +6),(0, 1, 0),(0, 0, 1))))
+            pos =  np.dot(R, np.dot(T, np.dot(R.I, v.T)))
+            ls = LineString([(self.xyz[0], self.xyz[1]), (float(pos[0]), float(pos[1]))])
+            intr = self.unsafe_location.intersection(ls)
+            if not intr.is_empty:
+                d = int(1000*(ls.length - intr.length))
+                if self.debug:
+                    print(self.sim_time, self.robot_name, "Virtual fence: decreasing front distance to %d" % d)
+                self.scan_distance_to_obstacle = min(self.scan_distance_to_obstacle, d)
+
+            left_angle = 0.787878
+            x,y = pol2cart(+6, left_angle)
+            Tl = np.asmatrix(np.array(((1, 0, x),(0, 1, y),(0, 0, 1))))
+            posL = np.dot(R, np.dot(Tl, np.dot(R.I, v.T)))
+            ls = LineString([(self.xyz[0], self.xyz[1]), (float(posL[0]), float(posL[1]))])
+            intr = self.unsafe_location.intersection(ls)
+            if not intr.is_empty:
+                d = int(1000*(ls.length - intr.length))
+                if self.debug:
+                    print(self.sim_time, self.robot_name, "Virtual fence: decreasing left distance to %d" % d)
+                self.scan_avg_distance_left = min(self.scan_avg_distance_left, d)
+
+            right_angle = -0.787878
```

Comment:
if the magic number is repeated I would rather use some constant

Reply:
Per above, this is not a magic constant in itself, instead it's derived from another magic constant.